lib/cfg.js

```diff
@@ -0,0 +1,368 @@
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright notice,
+//       this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+
+var _ = require('underscore-node');
+
+function ControlFlowGraph(rules){
+    this.rules = rules;
+    
+}
+
+function seperateCodeFromData(asmArr) {
+    
+    var isCode = function(asmArr, index){
+        return (((asmArr[index].source !== null) &&
+                (asmArr[index+1].source !== null)) ||
+                ((asmArr[index].source === null) &&
+                 (asmArr[index+1].source !== null)) ||
+                 ((asmArr[index].source !== null) &&
+                 (asmArr[index+1].source === null)));
+    };
+
+    var code = [];
+    var first = 0;
+    var last = asmArr.length;
+    
+    if(first == last) return code;
+
+    
+    while(first != last-1){
+        if(isCode(asmArr, first))
+            code.push(JSON.parse(JSON.stringify(asmArr[first])));
+        
+        ++first;
+    }
+    if(asmArr[first].source){
+        code.push(JSON.parse(JSON.stringify(asmArr[first])));
+    }
+
+    return code;
+}
+
+    var gccX86 = {
+                filterData:  seperateCodeFromData,
+                isFunctionEnd:function(x)  { return ((x[0] != ' ') && (x[0] != '.') &&
+                                                             (x.indexOf(':') != -1)) ;},
+
+                isBasicBlockEnd: function(x) { return x[0] == ".";} ,
+
+                getInstructionType:function(inst) {
+                                            inst = inst.trim();
+                                            if (inst === "") return -1;
+                                            else if(inst.includes("jmp")) return 0;
+                                            else if(inst[0] === 'j') return 1;
```

Comment:
Why does this not use `isJmpInstruction` ? Seems like it's repeating the same code in two places.

Reply:
the semantic of isJmpInstruction is to check for a all type of jmp instructions, I mean conditional ones and non conditional one (for ex: jmp). but here, even though it's the same code (`inst[0] === 'j'`), we are checking if it's a conditional jmp or  not, look at the previous test `inst.includes("jmp")` .

what I want to say here we are not testing for the same thing even though the code is the same. 

just to avoid confusion, conditional jmp here means je, jne, jg....

Reply:
Here I meant to refer to the (inst.includes("jmp") line - so:

```
if (inst === "") return -1
if (isJmpInstruction(inst)) return InstructionType_Jump;
if (inst[0] == "j") return InstructionType_JumpConditional;
```

I tagged the wrong line by mistake, sorry!

Reply:
this will make the third condition unreachable for conditional jmps.
Ah maybe the fact that I am writing Jmp in the name made you think that it's testing for `jmp .L4` but no, it returns true for all jmps either ways: conditional jmps or must taken jmp.